src/amplitude.js

```diff
@@ -137,146 +112,238 @@ Amplitude.prototype.init = function(apiKey, opt_userId, opt_config, callback) {
+ * @private
+ */
+var _parseConfig = function _parseConfig(options, config) {
+  if (type(config) !== 'object') {
+    return;
   }
+
+  // verifies config value is defined, is the correct type, and some additional value verification
+  var parseValidateLoad = function parseValidateLoad(key, expectedType) {
+    var input_value = config[key];
+    if(input_value === undefined || !utils.validateInput(input_value, key + ' option', expectedType)) {
+      return;
+    }
+    if (expectedType === 'boolean') {
+      options[key] = !!input_value;
+    } else {
+      options[key] = (expectedType === 'string' && !utils.isEmptyString(input_value) && input_value) ||
+                     (expectedType === 'number' && config[key] > 0 && input_value) ||
+                     options[key];
+    }
+   };
+
+   // the DEFAULT_OPTIONS object defines all valid keys, and provides expected types for the value
+   for (var key in DEFAULT_OPTIONS) {
+      if (DEFAULT_OPTIONS.hasOwnProperty(key)) {
+        parseValidateLoad(key, type(DEFAULT_OPTIONS[key]));
+      }
+   }
 };
 
+/**
+ * Run functions queued up by proxy loading snippet
+ * @private
+ */
 Amplitude.prototype.runQueuedFunctions = function () {
   for (var i = 0; i < this._q.length; i++) {
     var fn = this[this._q[i][0]];
-    if (fn && type(fn) === 'function') {
+    if (type(fn) === 'function') {
       fn.apply(this, this._q[i].slice(1));
     }
   }
   this._q = []; // clear function queue after running
 };
 
-Amplitude.prototype._apiKeySet = function(methodName) {
-  if (!this.options.apiKey) {
-    utils.log('apiKey cannot be undefined or null, set apiKey with init() before calling ' + methodName);
+/**
+ * Check that the apiKey is set before calling a function. Logs a warning message if not set.
+ * @private
+ */
+Amplitude.prototype._apiKeySet = function _apiKeySet(methodName) {
+  if (utils.isEmptyString(this.options.apiKey)) {
+    utils.log('Invalid apiKey. Please set a valid apiKey with init() before calling ' + methodName);
     return false;
   }
   return true;
 };
 
-Amplitude.prototype._loadSavedUnsentEvents = function(unsentKey) {
-  var savedUnsentEventsString = this._getFromStorage(localStorage, unsentKey);
-  if (savedUnsentEventsString) {
+/**
+ * Load saved events from localStorage. JSON deserializes event array. Handles case where string is corrupted.
+ * @private
+ */
+Amplitude.prototype._loadSavedUnsentEvents = function _loadSavedUnsentEvents(unsentKey) {
+  var savedUnsentEventsString = localStorage.getItem(unsentKey);
+  if (utils.isEmptyString(savedUnsentEventsString)) {
+    return []; // new app, does not have any saved events
+  }
+
+  if (type(savedUnsentEventsString) === 'string') {
     try {
-      return JSON.parse(savedUnsentEventsString);
-    } catch (e) {
-      // utils.log(e);
-    }
+      var events = JSON.parse(savedUnsentEventsString);
+      if (type(events) === 'array') { // handle case where JSON dumping of unsent events is corrupted
```

Comment:
It's supposed to deserialize into an array object, but if client's have broken JSON libraries, it deserializes into a string, which breaks the SDK